core/src/socket.rs

```diff
@@ -0,0 +1,239 @@
+                    if !success {
+                        continue;
+                    }
+                }
+                None => continue,
+
+                _ => {}
+            }
+
+            if let Some(received) = internal.poll() {
+                actions.push(SocketAction::Data(handle, received));
+            } else if matches!(is_connected, Some(false) | None) {
+                actions.push(SocketAction::Close(handle));
+                continue;
+            }
+
+            if let Some(to_send) = send_buffer.pop_front() {
+                internal.send(to_send);
+            }
+        }
+
+        if actions.is_empty() {
+            return;
+        }
+
+        let mut activation = Activation::from_nothing(context.reborrow());
+
+        for action in actions {
+            match action {
+                SocketAction::Connect(handle, success) => {
+                    let arena = &mut activation.context.sockets.0;
+                    let target = if success {
+                        arena
+                            .get(handle)
+                            .expect("only valid handles in SocketAction")
+                            .target
+                    } else {
+                        arena
+                            .remove(handle)
+                            .expect("only valid handles in SocketAction")
+                            .target
+                    };
+
+                    if success {
+                        let connect_evt =
+                            EventObject::bare_default_event(&mut activation.context, "connect");
+                        Avm2::dispatch_event(&mut activation.context, connect_evt, target);
+                    }
+                }
+                SocketAction::Close(handle) => {
+                    let target = activation
+                        .context
+                        .sockets
+                        .0
+                        .remove(handle)
+                        .expect("only valid handles in SocketAction")
+                        .target;
+
+                    let close_evt =
+                        EventObject::bare_default_event(&mut activation.context, "close");
+                    Avm2::dispatch_event(&mut activation.context, close_evt, target);
+                }
+                SocketAction::Data(handle, data) => {
+                    let target = activation
+                        .context
+                        .sockets
+                        .0
+                        .get(handle)
+                        .expect("only valid handles in SocketAction")
+                        .target;
+
+                    let socket = target.as_socket().expect("only SocketObjects in handles");
+                    socket.read_buffer().extend(data);
+
+                    let socket_data_evt =
+                        EventObject::bare_default_event(&mut activation.context, "socketData");
```

Comment:
This should be a `ProgressEvent`, not an `Event`. According to the Adobe docs:

Event Object Type: [flash.events.ProgressEvent](https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/flash/events/ProgressEvent.html)
property ProgressEvent.type = [flash.events.ProgressEvent.SOCKET_DATA](https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/flash/events/ProgressEvent.html#SOCKET_DATA)